server/instance_cloud_oauth.go

```diff
@@ -80,11 +83,37 @@ func (ci *cloudOAuthInstance) GetClient(connection *Connection) (Client, error)
 
 func (ci *cloudOAuthInstance) getClientForConnection(connection *Connection) (*jira.Client, *http.Client, error) {
 	oauth2Conf := ci.GetOAuthConfig()
-
 	ctx := context.Background()
 	tokenSource := oauth2Conf.TokenSource(ctx, connection.OAuth2Token)
 	client := oauth2.NewClient(ctx, tokenSource)
 
+	// Get a new token if Access Token has expired
+	if ci.Plugin.IsAccessTokenExpired(connection.OAuth2Token) {
+		ci.client.Log.Debug("AccessToken is about to expire (or) has expired. Refreshing the accessToken.")
+
+		// Set the AccessToken to empty in order for refresh to happen forcefully
+		token := connection.OAuth2Token
+		token.AccessToken = ""
+		tokenSource := oauth2Conf.TokenSource(ctx, token)
+
+		// Refresh the Token with its own mutex.
+		newToken, err := tokenSource.Token()
+		if err != nil {
+			ci.client.Log.Warn("Error while refreshing the Jira Access Token", "error", err)
+			return nil, nil, err
+		}
+
+		connection.OAuth2Token = newToken
+		tokenSource = oauth2Conf.TokenSource(ctx, newToken)
+		client = oauth2.NewClient(ctx, tokenSource)
+
+		// Store this new access token & refresh token to get a new access token in future when it has expired
+		err = ci.Plugin.userStore.StoreConnection(ci.Common().InstanceID, connection.MattermostUserID, connection)
+		if err != nil {
+			return nil, nil, err
+		}
```

Comment:
Seems odd that we are calling `TokenSource` twice, but I don't think this will cause issues. The access token will not need refresh the second time, so we won't be receiving a new refresh token, so this may work well.

Here's a snippet of what I was referring to in regards to explicitly checking if the refresh token changed. Also, note that in the Zoom plugin, the OAuth client is being instantiated by the `token` instead of the `tokenSource`. https://github.com/mattermost/mattermost-plugin-zoom/blob/fcdab70c5da49c704ac701abb55ac5ee3c2489d7/server/zoom/oauth.go#L158-L183

Reply:
This is resolved with the latest changes.